pkg/controller/provider/xavier/base/client.go

```diff
@@ -0,0 +1,335 @@
+		reply,
+		"result",
+		"execution-results",
+		"results",
+		"2",
+		"value",
+		"org.drools.core.runtime.rule.impl.FlatQueryResults",
+		"idResultMaps",
+		"element",
+		"0",
+		"element",
+		"0",
+		"value",
+		"org.jboss.xavier.analytics.pojo.output.workload.inventory.WorkloadInventoryReportModel")
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	providers, err := r.find(report, "recommendedTargetsIMS")
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	flags, err := r.find(report, "flagsIMS")
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	concerns := []string{}
+	if list, cast := flags.(List); cast {
+		for _, flag := range list {
+			concerns = append(concerns, flag.(string))
+		}
+	}
+	if len(concerns) == 0 {
+		recommended := false
+		if list, cast := providers.(List); cast {
+			for _, provider := range list {
+				if provider == "OCP" {
+					recommended = true
+					break
+				}
+			}
+		}
+		if !recommended {
+			concerns = append(
+				concerns,
+				"CNV not recommended.")
+		}
+	}
+	r.Adapter.UpdateConcerns(model, concerns)
+
+	Log.Info("Analyzed", "vm", model.String())
+
+	return
+}
+
+//
+// Build body
+func (r *Client) body(model libmodel.Model) (body Object) {
+	about, _ := r.about()
+	object := Object{
+		Provider:        "",
+		DataCenter:      "",
+		Cluster:         "",
+		VmName:          "",
+		DiskSpace:       0,
+		Memory:          0,
+		CpuCores:        0,
+		GuestOs:         "",
+		OsProduct:       "",
+		Product:         about.Product,
+		Version:         about.APIVersion,
+		HostName:        "",
+		HasMemoryHotAdd: false,
+		HasCpuHotAdd:    false,
+		HasCpuHotRemove: false,
+		HasCpuAffinity:  false,
+		HasDrsEnabled:   false,
+		HasHaEnabled:    false,
+		HasPassthrough:  false,
+		HasUsb:          false,
+		HasRdmDisk:      false,
+		BalloonedMemory: 0,
+		ScanDate:        time.Now().Unix(),
+	}
+	insert := Object{
+		"object": Object{
+			"org.jboss.xavier.analytics.pojo.input.workload.inventory.VMWorkloadInventoryModel": object,
+		},
+		"out-identifier": "input",
+	}
+	body = Object{
+		"lookup": "WorkloadInventoryKSession0",
+		"commands": List{
+			Object{
+				"insert": insert,
+			},
+			Object{
+				"fire-all-rules": Object{
+					"out-identifier": "firedActivations",
+				},
+			},
+			Object{
+				"query": Object{
+					"name":           "GetWorkloadInventoryReports",
+					"out-identifier": "WorkloadInventoryReports",
+					"arguments":      List{},
+				},
+			},
+		},
+	}
+	r.Adapter.UpdateBody(object, model)
+
+	return
+}
+
+//
+// Add headers.
+func (r Client) header() (header http.Header) {
+	header = http.Header{}
+	header.Add("Content-Type", "application/json")
+	encoded := base64.StdEncoding.EncodeToString(
+		[]byte(strings.Join([]string{
+			r.User,
+			r.Password,
+		}, ":")))
+	header.Add(
+		"Authorization",
+		"Basic "+encoded)
+
+	return
+}
+
+//
+// Post validation request.
+func (r *Client) post(in, out interface{}) (status int, err error) {
```

Comment:
I only see this method called in one place; is it necessary to use an interface type for these parameters? It looks like at the very least `out` will always be a `base.Object{}`.

Reply:
Yes, the post() can be narrowed to an `*Object` for both in and out.